File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]': 
        #Step 1:
        curr = head
        while curr:
            dummy = Node(curr.val)
            dummy.next = curr.next
            curr.next = dummy
            curr = curr.next.next
        
        #Step 2:
        curr = head
        while curr:
            fwd = curr.next.next
            dummy = curr.next
            dummy.random = curr.random.next if curr.random else None
            curr = fwd

        #Step 3:
        dummyNode = Node(-1)
        curr = head
        res = dummyNode

        while curr:
            res.next = curr.next
            res = res.next
            curr.next = res.next
            curr = res.next

        return dummyNode.next



        

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> map = new HashMap<>();
        Node curr = head;
        Node prev = null;
        Node newHead = null;

        while(curr != null){
            Node node = new Node(curr.val);
            if(newHead == null){
                newHead = node;
            }
            else{
                prev.next = node;
            }
            prev = node;
            map.put(curr, node);
            curr = curr.next;
        }

        curr = head;
        Node newCurr = newHead;

        while(curr != null){
            if(curr.random != null){
                newCurr.random = map.get(curr.random);
            }
            curr = curr.next;
            newCurr = newCurr.next;
        }
        return newHead;
    }
}